File: ha_kihul/video/animation.js
(function () {
    function $(id) { return document.getElementById(id); }

    function readFile(input, target, callback)
    {
        var file = input.files[0];

        if (file) {
            var reader = new FileReader();
            reader.onload = function(e) { target.setAttribute("src", e.target.result); }
            reader.readAsDataURL(file);
            callback();
        }
    }

    function time()
    {
        return Date.now() / 1000.0;
    }

    function keyToEventId(key)
    {
        return "event_" + String(key);
    }

    function getEvent(key)
    {
        var evt_id = keyToEventId(key);

        return events.hasOwnProperty(evt_id) ? events[evt_id] : null;
    }

    function loadSprite(parent_sprite, sprite)
    {
        var img = new Image(),
            id = sprites.length,
            children = [],
            // dbg_a = document.createElement("div"),
            evt, evt_id, effect_def, i, replace, dbg;

        if (sprite["position"].length !== 2) {
            return;
        }

        sprites.push({
            "image": img,
            // "dbg_a": dbg_a,
            "replace": null,
            "anchor": sprite["anchor"],
            "position": sprite["position"],
            "offset": [0, 0],
            "z": sprite["z"],
            "size": [0, 0],
            "center": [0, 0],
            "parent": parent_sprite,
            "rotate": 0,
            "scale": 1.0,
            "opacity": 1.0,
            "children": children
        });
        img.onload = function () {
            var s = sprites[id];

            s["size"] = [img.naturalWidth, img.naturalHeight];
            s["center"] = vscale(1.0 / 2.0, s["size"]);

            render();
        };
        img.src = sprite["name"];
        canvas.appendChild(img);
        // canvas.appendChild(dbg_a);

        for (i = 0; i < sprite["events"].length; ++i) {
            evt = sprite["events"][i];
            evt_id = keyToEventId(evt["key"]);

            if (!events.hasOwnProperty(evt_id)) {
                events[evt_id] = {
                    "state": "inactive",
                    "effects": []
                };
            }

            effect_def = evt["effect"];

            if (effect_def["type"] === "replace") {
                replace = new Image();
                replace.src = effect_def["img"];
                replace.style.position = "absolute";
                replace.style.left = "2000vw";
                canvas.appendChild(replace);
            } else {
                replace = null;
            }

            events[evt_id]["effects"].push(
                {
                    "def": effect_def,
                    "sprite": id,
                    "param": 0,
                    "replace": replace
                }
            );
        }

        for (i = 0; i < sprite["children"].length; ++i) {
            children.push(loadSprite(id, sprite["children"][i]));
        }

        return id;
    }

    function loadScene(scene)
    {
        loadSprite(null, scene["main"]);
        audio.src = scene["audio"];
    }

    function convert(pos)
    {
        var w = canvas_size[0],
            h = canvas_size[1],
            x = pos[0],
            y = pos[1];

        return [64.0 * x / w, 36.0 * y / h];
    }

    function vscale(scalar, vector)
    {
        return [scalar * vector[0], scalar * vector[1]];
    }

    function vsum(a, b)
    {
        return [a[0] + b[0], a[1] + b[1]];
    }

    function vdiff(a, b)
    {
        return [a[0] - b[0], a[1] - b[1]];
    }

    function rotate(point, origin, degrees)
    {
        var rad = (degrees * Math.PI) / 180,
            sin = Math.sin(rad),
            cos = Math.cos(rad),
            x, y;

        point = vdiff(point, origin);
        x = point[0];
        y = point[1];

        return vsum(origin, [x * cos - y * sin, x * sin + y * cos]);
    }

    function computeCenterRotationScaleOpacityOnScreen(sprite)
    {
        var center, rotation, scale, opacity,
            parent_sprite, parent_center, parent_rotation, parent_scale, parent_opacity,
            tmp;

        scale = sprite["scale"];
        opacity = sprite["opacity"];

        center = vsum(
            sprite["position"],
            vscale(scale, vdiff(sprite["center"], sprite["anchor"]))
        );

        rotation = sprite["rotate"];
        center = rotate(center, sprite["position"], rotation);
        center = vsum(center, vscale(sprite["scale"], sprite["offset"]));

        parent_sprite = sprite["parent"];

        if (parent_sprite !== null) {
            parent_sprite = sprites[parent_sprite];

            tmp = computeCenterRotationScaleOpacityOnScreen(parent_sprite);
            parent_center = tmp[0];
            parent_rotation = tmp[1];
            parent_scale = tmp[2];
            parent_opacity = tmp[3];
            center = vscale(parent_scale, center);
            center = vsum(parent_center, vdiff(center, vscale(parent_scale, parent_sprite["center"])));
            center = rotate(center, parent_center, parent_rotation);

            scale *= parent_scale;
            opacity *= parent_opacity;
            rotation += parent_rotation;
        }

        return [center, rotation, scale, opacity];
    }

    function render()
    {
        var i, l, sprite, img, size, topleft, style, anchor, center, rotation, scale, opacity, tmp;

        for (i = 0, l = sprites.length; i < l; ++i) {
            sprite = sprites[i];
            img = sprite["image"];
            style = img.style;
            tmp = computeCenterRotationScaleOpacityOnScreen(sprite);
            center = tmp[0];
            rotation = tmp[1];
            scale = tmp[2];
            opacity = tmp[3];

            if (sprite["replace"] !== null) {
                style.left = "2000vw";
                img = sprite["replace"];
                style = img.style;
            }

            size = [img.naturalWidth, img.naturalHeight];
            size = vscale(scale, size);
            topleft = convert(vdiff(center, vscale(1.0 / 2.0, size)));
            size = convert(size);

            style.position = "absolute";
            style.left = String(topleft[0]) + "vw";
            style.top = String(topleft[1]) + "vw";
            style.width = String(size[0]) + "vw";
            style.height = String(size[1]) + "vw";
            style.zIndex = sprite["z"];
            style.transform = "rotate(" + String(rotation) + "deg)";
            style.opacity = opacity;

            // sprite["dbg_a"].style.background = "#0000a0";
            // sprite["dbg_a"].style.position = "absolute";
            // sprite["dbg_a"].style.left = String(convert(center)[0]) + "vw";
            // sprite["dbg_a"].style.top = String(convert(center)[1]) + "vw";
            // sprite["dbg_a"].style.width = "5px";
            // sprite["dbg_a"].style.height = "5px";
            // sprite["dbg_a"].style.zIndex = "2000";
        }
    }

    function ease(t)
    {
        return t * t * (3 - 2 * t);
    }

    function animate()
    {
        var now = time(),
            done = true,
            delta, key, evt, state, param, effects, effect, duration, i, l, rewind_done, sprite, rec_ts, eased_param;

        if (previous_frame === null) {
            previous_frame = now;
        }

        if (playing) {
            timestamp.value = String(Math.floor((now - rec_started + rec_offset) * 1000) / 1000);
        }

        delta = now - previous_frame;
        param = 0;

        for (i = 0, l = sprites.length; i < l; ++i) {
            sprite = sprites[i];
            sprite["scale"] = 1.0;
            sprite["opacity"] = 1.0;
            sprite["rotate"] = 0.0;
            sprite["offset"] = [0.0, 0.0];
        }

        for (key in events) {
            if (!events.hasOwnProperty(key))
                continue;

            evt = events[key];
            state = evt["state"];

            if (state === "inactive")
                continue;

            effects = evt["effects"];
            rewind_done = true;

            for (i = 0, l = effects.length; i < l; ++i) {
                effect = effects[i];
                effect_def = effect["def"];

                if (effect_def.hasOwnProperty("duration")) {
                    param = effect["param"];

                    if (state === "started") {
                        duration = effect_def["duration"];

                        if (duration < 0.0001) {
                            param = 1;
                        } else {
                            param = Math.min(1, param + delta / duration);
                        }

                        if (param < 1) {
                            done = false;
                        }
                    } else {
                        duration = effect_def.hasOwnProperty("rewind")
                            ? effect_def["rewind"]
                            : effect_def["duration"];

                        if (duration < 0.0001) {
                            param = 0;
                        } else {
                            param = Math.max(0, param - delta / duration);
                        }

                        if (param > 0) {
                            done = false;
                            rewind_done = false;
                        }
                    }

                    effect["param"] = param;
                }

                sprite = sprites[effect["sprite"]];

                switch (effect_def["type"]) {
                    case "scale":
                        eased_param = ease(param);
                        sprite["scale"] *= eased_param * effect_def["scale"] + 1.0 - eased_param;
                        break;
                    case "opacity":
                        eased_param = ease(param);
                        sprite["opacity"] *= eased_param * effect_def["opacity"] + 1.0 - eased_param;
                        break;
                    case "rotate":
                        sprite["rotate"] += ease(param) * effect_def["angle"];
                        break;
                    case "translate":
                        sprite["offset"] = vsum(sprite["offset"], vscale(ease(param), effect_def["offset"]));
                        break;
                    case "replace":
                        if (state === "started") {
                            if (sprite["replace"] !== null) {
                                sprite["replace"].style.left = "2000vw";
                            }
                            sprite["replace"] = effect["replace"];
                        } else if (sprite["replace"] === effect["replace"]) {
                            sprite["replace"] = null;
                            effect["replace"].style.left = "2000vw";
                        }
                        break;
                }
            }

            if (state === "rewinding" && rewind_done) {
                evt["state"] = "inactive";
            }
        }

        render();

        if (done && (!playing)) {
            previous_frame = null;
        } else {
            previous_frame = now;
            rec_ts = now - rec_started + rec_offset;

            if (playing) {
                for (l = queued_events.length; (l > 0) && (queued_events[--l]["time"] <= rec_ts);) {
                    evt = queued_events.pop();

                    if (evt["type"] === "down") {
                        startEvent(evt["key"]);
                    } else {
                        rewindEvent(evt["key"]);
                    }
                }
            }

            requestAnimationFrame(animate);
        }
    }

    function rewindEvent(key)
    {
        var evt = getEvent(key);

        if (evt === null || evt["state"] !== "started")
            return false;

        evt["state"] = "rewinding";

        return true;
    }

    function startEvent(key)
    {
        var evt = getEvent(key);

        if (evt === null || evt["state"] === "started")
            return false;

        evt["state"] = "started";

        return true;
    }

    function recordEvent(evt_type, key)
    {
        var now;

        if (!recording) {
            return;
        }

        now = time();
        rec_events.push(
            {
                "type": evt_type,
                "key": key,
                "time": now - rec_started + rec_offset
            }
        );
    }

    function handleKeyUp(e)
    {
        var key = e["key"];

        if (!rewindEvent(key)) {
            return;
        }

        recordEvent("up", key);
        requestAnimationFrame(animate);
    }

    function handleKeyDown(e)
    {
        var key = e["key"],
            now;

        if (!startEvent(key)) {
            return;
        }

        recordEvent("down", key);
        requestAnimationFrame(animate);
    }

    function resetEffects()
    {
        var key, evt, i, l, sprite;

        for (key in events) {
            if (!events.hasOwnProperty(key))
                continue;

            evt = events[key];
            evt["state"] = "inactive";
            effects = evt["effects"];

            for (i = 0, l = effects.length; i < l; ++i) {
                effect = effects[i];

                if (effect.hasOwnProperty("param")) {
                    effect["param"] = 0;
                }
                if (effect["replace"] !== null) {
                    effect["replace"].style.left = "2000vw";
                }
            }
        }

        for (i = 0, l = sprites.length; i < l; ++i) {
            sprite = sprites[i];
            sprite["replace"] = null;
            sprite["rotate"] = 0;
            sprite["offset"] = [0, 0];
        }
    }

    function dumpEvents()
    {
        events_dump.value = (
            JSON.stringify(event_layers)
            .replace(/\{/g, "\n  {")
            .replace(/\],/g, "],\n")
        );
    }

    function stopPlaying()
    {
        if (playing && recording) {
            mergeRecordedEventsIntoSelectedLayer(time() - rec_started + rec_offset);
        }

        resetEffects();
        render();
        recording = false;
        playing = false;
        updatePlaybackButton();
        audio.pause();
        audio.currentTime = 0;
        dumpEvents();
    }

    function startPlaying()
    {
        var ex;

        if (events_dump.value) {
            try {
                event_layers = JSON.parse(events_dump.value);
            } catch (ex) {
                alert("Error: " + String(ex));
                return;
            }
        }

        resetEffects();
        render();
        play.innerHTML = "Stop";
        selected_layer = layer_selector.value;
        queued_events = queueEvents(record.checked);
        rec_offset = Number(timestamp.value);
        rec_events = [];
        audio.currentTime = rec_offset;
        rec_started = time();
        recording = record.checked;
        playing = true;
        audio.play();
        requestAnimationFrame(animate);
    }

    function togglePlayback()
    {
        if (playing) {
            stopPlaying();
        } else {
            startPlaying();
        }
    }

    function mergeRecordedEventsIntoSelectedLayer(rec_stopped)
    {
        var evts = [],
            layer = event_layers[selected_layer],
            evt, i, l;

        sortEventsDesc(layer);

        for (i = layer.length - 1; (i >= 0) && (layer[i]["time"] < rec_offset); --i) {
            evts.push(layer.pop());
        }

        for (i = 0, l = rec_events.length; i < l; ++i) {
            evts.push(rec_events[i]);
        }

        rec_events = [];

        for (i = layer.length - 1; (i >= 0) && (layer[i]["time"] <= rec_stopped); --i) {
            layer.pop();
        }

        for (i = layer.length - 1; i >= 0; --i) {
            evts.push(layer.pop());
        }

        sortEventsAsc(evts);
        event_layers[selected_layer] = evts;
    }

    function sortEventsAsc(evts)
    {
        evts.sort(function (a, b) { return a["time"] - b["time"]; });
    }

    function sortEventsDesc(evts)
    {
        evts.sort(function (a, b) { return b["time"] - a["time"]; });
    }

    function queueEvents(for_recording)
    {
        var q = [],
            layer, layer_id, i, j, l;

        for (i = 0; i < number_of_layers; ++i) {
            layer_id = "layer-" + String(i + 1);

            if (for_recording && layer_id === selected_layer) {
                continue;
            }

            layer = event_layers[layer_id];

            for (j = 0, l = layer.length; j < l; ++j) {
                q.push(layer[j]);
            }
        }

        sortEventsDesc(q);

        return q;
    }

    function updatePlaybackButton()
    {
        play.innerHTML = record.checked ? "Record" : "Play";
    }

    var body = document.getElementsByTagName("body")[0],
        record = $("record"),
        play = $("play"),
        recording = false,
        playing = false,
        rec_started = 0,
        rec_offset = 0,
        rec_events = [],
        canvas = $("canvas"),
        audio = $("audio"),
        layer_selector = $("layer"),
        selected_layer = null,
        scene = window.scene,
        timestamp = $("timestamp"),
        sprites = [],
        canvas_size = window.scene["size"],
        events = {},
        event_layers = {},
        number_of_layers = 20,
        queued_events = null,
        events_dump = $("events"),
        previous_frame = null;

    function main()
    {
        var i, l, option, layer_id;

        for (i = 0; i < number_of_layers; ++i) {
            layer_id = "layer-" + String(i + 1);
            option = document.createElement("option");
            option.value = layer_id;
            option.innerHTML = String(i + 1);
            layer_selector.appendChild(option);
            event_layers[layer_id] = [];
        }

        dumpEvents();

        $("form").onsubmit = function () { return false; }
        play.onclick = togglePlayback;
        record.onchange = updatePlaybackButton;

        updatePlaybackButton();

        body.onkeyup = handleKeyUp;
        body.onkeydown = handleKeyDown;

        loadScene(scene);

        animate();
    }

    main();
})();

File: ha_kihul/video/animation.py
#!/usr/bin/python3

import sys
import os
import os.path
import re
import math
import argparse
import json
import time

from fractions import Fraction

from PIL import Image, ImageDraw, ImageMath


def vscale(scalar, vector):
    return (scalar * vector[0], scalar * vector[1])


def vsum(a, b):
    return (a[0] + b[0], a[1] + b[1])


def vdiff(a, b):
    return (a[0] - b[0], a[1] - b[1])


def rotate(point, origin, degrees):
    rad = (degrees * math.pi) / 180
    sin = math.sin(rad)
    cos = math.cos(rad)
    x, y = vdiff(point, origin);

    return vsum(origin, (x * cos - y * sin, x * sin + y * cos))


def ease(t):
    return t * t * (3 - 2 * t)


def sort_events_desc(evts):
    return sorted(evts, key=lambda evt: evt["time"], reverse=True)


class Animation:
    def __init__(self):
        self.sprites = []
        self.sprites_by_z = []
        self.events = {}
        self.now = 0
        self.previous_frame = None
        self.queued_events = None

    def key_to_event_id(self, key):
        return f"event_{key}"

    def get_event(self, key):
        return self.events.get(self.key_to_event_id(key), None)

    def load_sprite(self, parent_sprite, sprite):
        if len(sprite["position"]) != 2:
            return;

        img = Image.open(sprite["name"])

        # d = ImageDraw.Draw(img)
        # d.rectangle(((0, 0), img.size), outline=(255, 255, 0), width=5)

        sid = len(self.sprites)
        children = []

        self.sprites.append({
            "image": img,
            "center": vscale(0.5, img.size),
            "replace": None,
            "anchor": tuple(sprite["anchor"]),
            "position": tuple(sprite["position"]),
            "offset": (0, 0),
            "z": sprite["z"],
            "size": img.size,
            "scale": 1.0,
            "opacity": 1.0,
            "parent": parent_sprite,
            "rotate": 0,
            "children": children
        })

        for evt in sprite["events"]:
            evt_id = self.key_to_event_id(evt["key"]);

            if evt_id not in self.events:
                self.events[evt_id] = {
                    "state": "inactive",
                    "effects": [],
                }

            effect_def = evt["effect"];

            if effect_def["type"] == "replace":
                replace = Image.open(effect_def["img"])
            else:
                replace = None

            self.events[evt_id]["effects"].append({
                "def": effect_def,
                "sprite": sid,
                "param": 0,
                "replace": replace
            })

        for child in sprite["children"]:
            children.append(self.load_sprite(sid, child))

        return sid

    def load_scene(self, scene):
        self.load_sprite(None, scene["main"])

    def compute_center_rotation_scale_opacity_on_screen(self, sprite):
        rotation = sprite["rotate"];
        position = sprite["position"]
        scale = sprite["scale"]
        opacity = sprite["opacity"]
        center = vsum(
            position,
            vscale(scale, vdiff(sprite["center"], sprite["anchor"]))
        )
        center = rotate(center, position, rotation)
        center = vsum(center, vscale(sprite["scale"], sprite["offset"]))
        parent_sprite = sprite["parent"]

        if parent_sprite is not None:
            parent_sprite = self.sprites[parent_sprite]
            parent_center, parent_rotation, parent_scale, parent_opacity = self.compute_center_rotation_scale_opacity_on_screen(parent_sprite)
            center = vscale(parent_scale, center)
            center = vsum(parent_center, vdiff(center, vscale(parent_scale, parent_sprite["center"])))
            center = rotate(center, parent_center, parent_rotation)

            scale *= parent_scale
            opacity *= parent_opacity
            rotation += parent_rotation

        return (center, rotation, scale, opacity)

    def render(self, out_img):
        for sid in self.sprites_by_z:
            sprite = self.sprites[sid]
            img = sprite["image"]

            if sprite["replace"] is not None:
                img = sprite["replace"]

            center, rotation, scale, opacity = self.compute_center_rotation_scale_opacity_on_screen(sprite)

            if opacity < 0.0001:
                continue

            resample = Image.BICUBIC
            # resample = Image.NEAREST
            size = vscale(scale, img.size)
            img = img.resize((int(size[0]), int(size[1])), resample)
            img = img.rotate(-rotation, resample, True)

            if opacity < 1.0:
                opacity = max(0.0, opacity)
                alpha = img.getchannel("A")
                alpha = ImageMath.eval("convert(float(a) * {} + 0.5, 'L')".format(opacity), a=alpha)
                img.putalpha(alpha)

            dest_top_left = vdiff(center, vscale(1.0 / 2.0, img.size))
            dest_top_left = [int(dest_top_left[0]), int(dest_top_left[1])]
            src_top_left = [0, 0]

            if dest_top_left[0] < 0:
                src_top_left[0] = - dest_top_left[0]
                dest_top_left[0] = 0

            if dest_top_left[1] < 0:
                src_top_left[1] = - dest_top_left[1]
                dest_top_left[1] = 0

            out_img.alpha_composite(img, tuple(dest_top_left), tuple(src_top_left))

            # d = ImageDraw.Draw(out_img)
            # d.rectangle((top_left, vsum(top_left, (10, 10))), (255, 0, 128))

    def animate(self):
        done = True

        if self.previous_frame is None:
            self.previous_frame = self.now

        delta = self.now - self.previous_frame

        while len(self.queued_events) > 0 and self.queued_events[-1]["time"] <= self.now:
            evt = self.queued_events.pop()

            if evt["type"] == "down":
                self.start_event(evt["key"])
            else:
                self.rewind_event(evt["key"])

        for sprite in self.sprites:
            sprite["scale"] = 1.0
            sprite["opacity"] = 1.0
            sprite["rotate"] = 0.0
            sprite["offset"] = (0.0, 0.0)

        for evt in self.events.values():
            state = evt["state"]

            if state == "inactive":
                continue;

            effects = evt["effects"]
            rewind_done = True

            for effect in effects:
                effect_def = effect["def"]

                if "duration" in effect_def:
                    param = effect["param"]

                    if state == "started":
                        duration = effect_def["duration"]

                        if duration < 0.0001:
                            param = 1.0
                        else:
                            param = min(1.0, param + delta / duration)

                        if param < 1.0:
                            done = False
                    else:
                        if "rewind" in effect_def:
                            duration = effect_def["rewind"]
                        else:
                            duration = effect_def["duration"]

                        if duration < 0.0001:
                            param = 0.0
                        else:
                            param = max(0.0, param - delta / duration)

                        if param > 0.0:
                            done = False
                            rewind_done = False

                    effect["param"] = param

                sprite = self.sprites[effect["sprite"]]

                if effect_def["type"] == "scale":
                    eased_param = ease(param)
                    sprite["scale"] *= eased_param * effect_def["scale"] + 1.0 - eased_param;

                elif effect_def["type"] == "opacity":
                    eased_param = ease(param)
                    sprite["opacity"] *= eased_param * effect_def["opacity"] + 1.0 - eased_param;

                elif effect_def["type"] == "rotate":
                    sprite["rotate"] += ease(param) * effect_def["angle"]

                elif effect_def["type"] == "translate":
                    sprite["offset"] = vsum(
                        sprite["offset"],
                        vscale(ease(param), effect_def["offset"])
                    )

                elif effect_def["type"] == "replace":

                    if state == "started":
                        sprite["replace"] = effect["replace"]

                    elif sprite["replace"] is effect["replace"]:
                        sprite["replace"] = None

            if state == "rewinding" and rewind_done:
                evt["state"] = "inactive"

        self.previous_frame = self.now

    def rewind_event(self, key):
        evt = self.get_event(key)

        if evt is None or evt["state"] != "started":
            return

        evt["state"] = "rewinding"

    def start_event(self, key):
        evt = self.get_event(key);

        if evt is None or evt["state"] == "started":
            return

        evt["state"] = "started"

    def queue_events(self, event_layers):
        q = []

        for layer in event_layers.values():
            for evt in layer:
                q.append(evt)

        return sort_events_desc(q)

    def main(self, argv):
        if len(argv) < 4:
            print(f"Usage: {argv[0]} scene.js events.js FPS [end]", file=sys.stderr)
            return 1

        with open(argv[1], "r") as f:
            scene = json.loads("{" + f.read().split("{", 1)[1].rsplit("}", 1)[0] + "}")

        with open(argv[2], "r") as f:
            event_layers = json.loads(f.read())

        fps = max(1.0, min(60.0, float(argv[3])))

        end = None

        if len(argv) > 4:
            end = float(argv[4])

        self.load_scene(scene)
        self.queued_events = self.queue_events(event_layers)

        self.sprites_by_z = sorted(
            range(0, len(self.sprites)),
            key=lambda s: self.sprites[s]["z"]
        )

        if len(self.queued_events) < 1:
            return 0

        last_evt = self.queued_events[0]
        evt = self.get_event(last_evt["key"])

        if end is None:
            end = (
                last_evt["time"]
                + max(
                    effect["def"]["duration"] + effect["def"].get("rewind", effect["def"]["duration"])
                    for effect in evt["effects"] if "duration" in effect["def"]
                )
            ) + 2.0

        img = Image.new("RGBA", tuple(scene["size"]), (255, 255, 255))
        canvas = ImageDraw.Draw(img)
        stdout = os.fdopen(sys.stdout.fileno(), 'wb')
        self.now = 0
        prev_percent = ""
        start = time.time()

        while self.now < end:
            percent = self.format_number((100.0 * self.now) / end)

            if percent != prev_percent:
                eta = self.compute_eta(start, self.now, end)
                print(
                    f"Rendering... {percent}%    {self.format_number(self.now)}s/{self.format_number(end)}s    ETA: {eta}    ",
                    file=sys.stderr
                )
                prev_percent = percent

            canvas.rectangle(((0, 0), tuple(scene["size"])), (0, 255, 0))
            self.animate()
            self.render(img)
            img.save(stdout, "BMP")
            self.now += 1.0 / fps

        return 0

    def format_number(self, n):
        return str(int(n * 100.0 + 0.5) / 100.0)

    def compute_eta(self, start, ready, end):
        if ready < 0.01:
            return "???"

        now = time.time()
        elapsed = now - start
        backlog = end - ready

        return self.format_time((backlog * elapsed) / ready)

    def format_time(self, t):
        t = int(t + 0.5)

        return ":".join([
            str(int(t / 3600)),
            self.format_time_component(int((t % 3600) / 60)),
            self.format_time_component(t % 60),
        ])

    def format_time_component(self, c):
        if c > 9:
            return str(c)

        return "0" + str(c)


sys.exit(Animation().main(sys.argv));
